File: python/599.py
"""Project Euler Problem 599: Distinct colorings of a 2x2x2 Rubik's Cube.

Find the number of distinct colorings of a 2x2x2 Rubik's Cube using N colors.

The number of corners with 3 colors is N(N-1)(N-2) divided by the 3 ways to
rotate it, the number of corners with 2 colors is N (the doubled color)
times N-1 (the single color), and the number of corners with a single color
is simply N. The number of ways to choose the 8 corners of a 2x2x2 can be
computed with balls and urns. However, the orientation of the last corner
of a 2x2x2 is fixed. This means that if all corners are multicolored (have
at least two colors), then there are 2 additional distinct ways of orienting
the last corner.
"""

from __future__ import annotations

from math import comb


def solve() -> int:
    """Solve Problem 599."""
    N = 10

    num_multicolored_corners = N * (N - 1) * (N - 2) // 3 + N * (N - 1)
    num_corners = num_multicolored_corners + N

    ans = comb(num_corners + 7, 8) + 2 * comb(num_multicolored_corners + 7, 8)

    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/599.c
/*
 * Project Euler Problem 599: Distinct colorings of a 2x2x2 Rubik's Cube.
 *
 * The number of distinct colorings using N colors.
 */

#include <stdio.h>
#include <stdint.h>

typedef long long ll;

/* Compute C(n, k) */
ll comb(int n, int k) {
    if (k < 0 || k > n) return 0;
    if (k == 0 || k == n) return 1;
    if (k > n - k) k = n - k;
    ll result = 1;
    for (int i = 0; i < k; i++) {
        result = result * (n - i) / (i + 1);
    }
    return result;
}

int main() {
    int N = 10;

    ll num_multicolored_corners = (ll)N * (N - 1) * (N - 2) / 3 + (ll)N * (N - 1);
    ll num_corners = num_multicolored_corners + N;

    ll ans = comb((int)(num_corners + 7), 8) + 2 * comb((int)(num_multicolored_corners + 7), 8);

    printf("%lld\n", ans);
    return 0;
}
